In [1]:
# install line profiler
!pip install line_profiler

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 7.3 MB/s  0:00:00 eta 0:00:01


In [5]:
# magics: ensures that any changes to the modules loaded below will be re-loaded automatically
%load_ext autoreload
%autoreload 2
%load_ext line_profiler

# load general packages
import numpy as np

# load modules related to this exercise
from model_zucher_exante import zurcher
from Solve_NFXP_exante import solve_NFXP
import estimate_NFXP_exante as estimate

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload
The line_profiler extension is already loaded. To reload it, use:
  %reload_ext line_profiler


[autoreload of model_zucher_exante failed: Traceback (most recent call last):
  File "/home/marius/data/0_Programme/anaconda3/lib/python3.13/site-packages/IPython/extensions/autoreload.py", line 325, in check
    superreload(m, reload, self.old_objects)
    ~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/marius/data/0_Programme/anaconda3/lib/python3.13/site-packages/IPython/extensions/autoreload.py", line 580, in superreload
    module = reload(module)
  File "/home/marius/data/0_Programme/anaconda3/lib/python3.13/importlib/__init__.py", line 129, in reload
    _bootstrap._exec(spec, module)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 866, in _exec
  File "<frozen importlib._bootstrap_external>", line 1023, in exec_module
  File "<frozen importlib._bootstrap_external>", line 1161, in get_code
  File "<frozen importlib._bootstrap_external>", line 1091, in source_to_code
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_remo

# Exercise 1

#### 1. Look at ReadMe.txt to get an overview of the code

#### 2. Invistigate how the code works, that is ensure you understand:
<il type ="a">
<li> zurcher.init</li>
<li> zurcher.setup</li>
<li> zurcher.create_grid</li>
<li> zucher.state_transition </li>
<li> zucher.bellman </li>

You can see how they are called below
    

#### 3. Fill in the missing stuff in the function zucher.bellman and run the code below

In [6]:
do_settings = {
    'RC': 0.5,
    'n': 12,
    'p':[0.65,0.2,0.1]   
}
model = zurcher(**do_settings)

print('Model grid:\n',model.grid)
print('Transition probabilities conditional on not replacing:\n',model.P1)
print('Transition probabilities conditional on replacing:\n',model.P2)
ev,pk, dev = model.bellman(np.zeros((model.n)),output=3)
print('Bellman one run:\n',ev)



Model grid:
 [ 0  1  2  3  4  5  6  7  8  9 10 11]
Transition probabilities conditional on not replacing:
 [[0.65 0.2  0.1  0.05 0.   0.   0.   0.   0.   0.   0.   0.  ]
 [0.   0.65 0.2  0.1  0.05 0.   0.   0.   0.   0.   0.   0.  ]
 [0.   0.   0.65 0.2  0.1  0.05 0.   0.   0.   0.   0.   0.  ]
 [0.   0.   0.   0.65 0.2  0.1  0.05 0.   0.   0.   0.   0.  ]
 [0.   0.   0.   0.   0.65 0.2  0.1  0.05 0.   0.   0.   0.  ]
 [0.   0.   0.   0.   0.   0.65 0.2  0.1  0.05 0.   0.   0.  ]
 [0.   0.   0.   0.   0.   0.   0.65 0.2  0.1  0.05 0.   0.  ]
 [0.   0.   0.   0.   0.   0.   0.   0.65 0.2  0.1  0.05 0.  ]
 [0.   0.   0.   0.   0.   0.   0.   0.   0.65 0.2  0.1  0.05]
 [0.   0.   0.   0.   0.   0.   0.   0.   0.   0.65 0.2  0.15]
 [0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.65 0.35]
 [0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   0.   1.  ]]
Transition probabilities conditional on replacing:
 [[0.65 0.2  0.1  0.05 0.   0.   0.   0.   0.   0.   0.   0.  ]
 [0.65 0.2  0.1  0.05

    
### 4. Solve the model. In order to solve the model, you should understand:
<li> solve_NFXP.init</li>
<li> solve_NFXP.setup</li>
<li> solve_NFXP.poly </li>
<li> solve_NFXP.sa </li>
<li> solve_NFXP.nk </li>
</il>
You can see how they are called below: 

In [7]:
algorithm = 'poly'
do_settings_solver = {
    'sa_min': 10,
    'sa_max': 20,  
    'printfxp': 2
}

solver = solve_NFXP(**do_settings_solver)
model = zurcher()

if algorithm == 'sa':
    ev = solver.sa(model.bellman)
if algorithm == 'poly':
    ev = solver.poly(model.bellman)
else:
    print('Algorithm must be "sa" or "poly"')



Begin contraction iterations (for the 1 time)
Iteration 1, tol     0.4273, tol(j)/tol(j-1)          1
Iteration 2, tol     0.4272, tol(j)/tol(j-1)     0.9999
Iteration 3, tol     0.4272, tol(j)/tol(j-1)     0.9999
Iteration 4, tol     0.4272, tol(j)/tol(j-1)     0.9999
Iteration 5, tol     0.4271, tol(j)/tol(j-1)     0.9999
Iteration 6, tol     0.4271, tol(j)/tol(j-1)     0.9999
Iteration 7, tol      0.427, tol(j)/tol(j-1)     0.9999
Iteration 8, tol      0.427, tol(j)/tol(j-1)     0.9999
Iteration 9, tol     0.4269, tol(j)/tol(j-1)     0.9999
Iteration 10, tol     0.4269, tol(j)/tol(j-1)     0.9999
Iteration 11, tol     0.4269, tol(j)/tol(j-1)     0.9999
Iteration 12, tol     0.4268, tol(j)/tol(j-1)     0.9999
Iteration 13, tol     0.4268, tol(j)/tol(j-1)     0.9999
Iteration 14, tol     0.4267, tol(j)/tol(j-1)     0.9999
Iteration 15, tol     0.4267, tol(j)/tol(j-1)     0.9999
Iteration 16, tol     0.4266, tol(j)/tol(j-1)     0.9999
Iteration 17, tol     0.4266, tol(j)/tol(j-1)     0

#### 5. Now we have to estimate the model. In order to estimate the model, you should understand:
<il type ="a">
<li> zurcher.read_busdata </li>
<li> estimate_NFXP.estimate  </li>
<li> estimate_NFXP.ll  </li>
</il>

You can see how they are called below:

#### 6. Fill in the missing stuff in the function estimate_NFXP.ll, and estimate the model to check that your results are correct
##### Check for missing sign somewhere, c should be positive

In [9]:
# Set up the model
model = zurcher()

# Set-up solver
solver = solve_NFXP()

# Read the data
data = model.read_busdata(bustypes=[1,2,3,4])
samplesize = data.shape[0]

# Estimate the model
import time
t0 = time.time()
theta0 = [0,0]

# args for nfxp estimate
nfxp_model, optim_res, pnames, theta_hat, Avar, converged=estimate.estimate(model, solver,data,theta0=theta0, twostep=0)

t1 = time.time()
time = t1-t0

# Print the result
print(f'Structual estimation using busdata from Rust(1987)')
print(f'Beta        = {model.beta:.4f}')
print(f'n           = {model.n}')
print(f'Sample size = {samplesize}\n \n')

print(f'Parameters     Estimates    s.e. ') 
print(f'{pnames[0]}             {theta_hat[0]:.4f}     {np.sqrt(Avar[0,0]):.4f} ')
print(f'{pnames[1]}              {theta_hat[1]:.4f}     {np.sqrt(Avar[1,1]):.4f} \n ')
print(f'{pnames[2]}(1)           {theta_hat[2]:.4f}     {np.sqrt(Avar[2,2]):.4f}  ')
print(f'{pnames[2]}(2)           {theta_hat[3]:.4f}     {np.sqrt(Avar[3,3]):.4f}  ')
print(f'{pnames[2]}(3)           {theta_hat[4]:.4f}     {np.sqrt(Avar[4,4]):.4f}  ')
print(f'{pnames[2]}(4)           {theta_hat[5]:.4f}     {np.sqrt(Avar[5,5]):.4f}  \n')


print(f'Log-likelihood {-optim_res.fun*samplesize:.2f}') 
print(f'runtime (seconds) {time:.4f}')
print(f'The model converged: {converged}')

Structual estimation using busdata from Rust(1987)
Beta        = 0.9999
n           = 175
Sample size = 8156
 

Parameters     Estimates    s.e. 
RC             4.7610     0.6378 
c              -1.3001     0.1068 
 
p(1)           0.1069     0.0034  
p(2)           0.5154     0.0055  
p(3)           0.3621     0.0053  
p(4)           0.0143     0.0013  

Log-likelihood -8660.55
runtime (seconds) 0.5429
The model converged: True


#### 7. Try using line_profiler in python. This gives you a lot of information about the performance of your code

In [10]:
%lprun -f estimate.ll  -f estimate.estimate estimate.estimate(model, solver,data,theta0=theta0, twostep=0)

Timer unit: 1e-09 s

Total time: 0.485592 s
File: /home/marius/Data/0_Programme/Github_desktop_repos/2026_structural_metrics_coursework/lab4/estimate_NFXP_exante.py
Function: ll at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
     1                                           #  NFXP class for structural estimation of discrete choice models.
     2                                           
     3        70    8187294.0 116961.3      1.7  import numpy as np

Total time: 0.514935 s
File: /home/marius/Data/0_Programme/Github_desktop_repos/2026_structural_metrics_coursework/lab4/estimate_NFXP_exante.py
Function: estimate at line 11

Line #      Hits         Time  Per Hit   % Time  Line Contents
    11                                           
    12                                           def estimate(model,solver,data,theta0=[0,0],twostep=0): # theta0 = initial value of structural parameter
    13         1      13759.0  13759.0      0.0      global ev
    14  

In [11]:
%lprun -f solve_NFXP.nk -f solve_NFXP.poly solve_NFXP.poly(solver,model.bellman)

Timer unit: 1e-09 s

Total time: 0.0162548 s
File: /home/marius/Data/0_Programme/Github_desktop_repos/2026_structural_metrics_coursework/lab4/Solve_NFXP_exante.py
Function: solve_NFXP.poly at line 31

Line #      Hits         Time  Per Hit   % Time  Line Contents
    31                                               def poly(self,bellman, V0=np.zeros(1), beta= 0.0, output=1):
    32                                           
    33         1       6146.0   6146.0      0.0          t0poly = time.time()  # set the starting time
    34                                           
    35         6      15155.0   2525.8      0.1          for k in range(self.max_fxpiter):
    36                                           
    37                                                       # 1. CONTRACTION ITERATIONS (S-A)
    38         5      10964.0   2192.8      0.1              if self.printfxp>0:
    39                                                           print(f'Begin contraction iterations 

a) Now try changing the optimizer options, and turn the use of the non-numerical Hessian off . What happens?

b) Now also try it with the analytical gradient off, what happens?

#### 8. Try estimate the model for different values of $\beta$. 

(a) Why can we not estimate $\beta$?

(b) When estimating with different $\beta$, do the changes in the estimates of c and/or RC make intuitively sense?

(c) Can you think of some data/variation, which could allow us to identify $\beta$?

#### 9. We use the latest EV guess to start the solve-procedure even though we change $\theta$ from one likelihood iteration to another. Why do you think we do that? 
(a) What if we started over with EV=0 each iteration? Try that and see what happens with the parameters and the numerical performance.

#### 10. Try setting the maximum number of miles (odometer reading) to 900. Now the absorbing state is much higher. 

(a) If we adjust the number of grid points as well, so that we have a comparable model (multiply the number of grids by 2), do we get a better fit? 

(b) Try to lower the number of grid points to 175 again. How do the parameters change? Are the changes intuitive? 

(c) What if you change the max to 225 and half the number of grids (hint: what goes wrong?)?